### PR TITLE
Add a collectsub client type Datasource for collectors

### PR DESCRIPTION
**Pull request by lumjjb**
Add a collectsub client type Datasource for collectors to use as part of #244 

Depends on https://github.com/guacsec/guac/pull/402

### CODE REVIEW

File: pkg/collectsub/client/client.go [hunk 1]
```diff
@@ -25,30 +25,36 @@ import (
...
+	Close()
+}
+
+type client struct {
 	client pb.ColectSubscriberServiceClient
 	conn   *grpc.ClientConn
 }
 
-func NewClient(addr string) (*Client, error) {
+func NewClient(addr string) (Client, error) {
```
Comment (naveensrinivasan) [inline pkg/collectsub/client/client.go:39]: Is it possible to return a pointer to the `*Client` instead of the `Client` itself?

File: pkg/collectsub/datasource/csubsource/csubsource.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// DataSourcesUpdate will return a channel which will get an element
+// if the CollectSource has new data. Upon update, nil is inserted
+// into the channel and non-nil if the channel no longer is able to
+// serve updates.
+func (d *csubDataSources) DataSourcesUpdate(ctx context.Context) (<-chan error, error) {
+	updateChan := make(chan error)
+	go func() {
+		for {
+			select {
+			case <-time.After(d.pollDuration):
```
Comment (naveensrinivasan) [inline pkg/collectsub/datasource/csubsource/csubsource.go:71]: I would recommend using `ticker := time.NewTicker(d.pollDuration)`
The use of a ticker instead of time.After can improve the reliability of the code in several ways:

1. A ticker is a periodic timer that repeatedly sends a value on a channel, while time.After sends a value only once after a specified duration. This means that a ticker can be used to run a task at regular intervals, while time.After would only trigger once after the specified duration.

1. The ticker's channel can be stopped or reset, providing more control over the timing of the periodic task. With time.After, once the specified duration has passed, there is no way to stop or reset the timer.

1. Tickers are more efficient than using time.After in a loop, as they use a single underlying timer. Each time time.After is used, a new timer is created, leading to increased resource usage.

In short, tickers provide a more flexible and efficient way to perform periodic tasks compared to using time.After.

File: pkg/collectsub/datasource/csubsource/csubsource.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			d.GitDataSources = append(d.GitDataSources, datasource.Source{
+				Value: e.Value,
+			})
+		case pb.CollectDataType_DATATYPE_OCI:
+			d.OciDataSources = append(d.OciDataSources, datasource.Source{
+				Value: e.Value,
+			})
+
+		default:
+			// unhandled datatype, skip
```
Comment (naveensrinivasan) [inline pkg/collectsub/datasource/csubsource/csubsource.go:109]: A log of unhandled would be helpful to debug.

File: pkg/collectsub/datasource/filesource/filesource.go [hunk 4]
```diff
@@ -115,6 +116,12 @@ func (d *fileDataSources) DataSourcesUpdate() (<-chan error, error) {
 				updateChan <- err
 				fmt.Printf("got err while watching: %+v\n", err)
 				return
+
+			case <-ctx.Done():
+				err := fmt.Errorf("file watcher ending from context closure")
+				updateChan <- err
+				fmt.Printf("ctx is closed: %+v\n", err)
```
Comment (naveensrinivasan) [inline pkg/collectsub/datasource/filesource/filesource.go:123]: Recommend to use log instead of fmt.

File: pkg/collectsub/datasource/csubsource/csubsource.go (hunk 2 above)
Comment (lumjjb) [reply pkg/collectsub/datasource/csubsource/csubsource.go:71]: hmm yea that seems suitable! I'm guessing probably a ticker + do a reset after each select is completed so that the ticker doesn't get backlogged if the logic takes too long. Learn something new everyday. Thanks!

File: pkg/collectsub/datasource/filesource/filesource.go (hunk 4 above)
Comment (lumjjb) [reply pkg/collectsub/datasource/filesource/filesource.go:123]: +1 , will remove these statements since the errors are propagated and would be handled elsewhere

Comment (naveensrinivasan) [reply pkg/collectsub/datasource/filesource/filesource.go:123]: 👍

File: pkg/collectsub/datasource/csubsource/csubsource.go (hunk 2 above)
Comment (naveensrinivasan) [reply pkg/collectsub/datasource/csubsource/csubsource.go:71]: 👍

File: pkg/collectsub/client/client.go (hunk 1 above)
Comment (lumjjb) [reply pkg/collectsub/client/client.go:39]: it is an interface now so it is already a pointer

File: pkg/collectsub/datasource/csubsource/csubsource_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		t.Errorf("unable to create FileDataSources: %v", err)
+		return
+	}
+
+	upChan, err := cds.DataSourcesUpdate(ctx)
+	if err != nil {
+		t.Errorf("unable to get DataSourcesUpdate: %v", err)
+	}
+
+	ds, err := cds.GetDataSources(ctx)
```
Comment (pxp928) [inline pkg/collectsub/datasource/csubsource/csubsource_test.go:106]: is there a reason why `GetDataSources` gets called twice and compared in this test?

**Review by pxp928 — APPROVED** [review]
LGTM! Question about one of the tests.

File: pkg/collectsub/datasource/csubsource/csubsource_test.go (hunk 5 above)
Comment (lumjjb) [reply pkg/collectsub/datasource/csubsource/csubsource_test.go:106]: yup - it gets updated and it checks to see if it got the updated content.

**Review by mihaimaruseac — APPROVED** [review]
